#' @title metadata MT to fasta sequences files
#'
#' @description
#' Generate and saved mutated sequences as fasta files.
#' The mutated sequences are generated by random calling of nucleotides based on nucleotide frequencies of the 1Mb sequences.
#'
#' @param DNAstring <DNAString> : the sequence of the chromosome, it must be the same sequence that was use to generate the WildType metadata dataFrame (with genDFWT or genDFWTstep2)
#' @param metadataWT <dataFrame> : a dataframe generated by metadataWT_seqWide or metadataWT_customWindow.
#' @param metadataMT <dataFrame> : a dataframe generated by metadataMT.
#' @param workdir <character> : the path to the folder that will contain a "fasta" directory with all the fasta sequences. The folder are created if they doesn't exist.
#' @param gzip <boolean> : if TRUE, the fasta files are saved in gzip format, otherwise they are saved in plain text format.
#'
#' @return <message> : a message indicating the number of fasta files saved and the output directory.
#'
#' @importFrom Biostrings writeXStringSet DNAStringSet
#' @importFrom OSMAdev shufDNA
#' @export

writeFastaMT = function(DNAstring, metadataWT, metadataMT, workdir = "./", gzip = TRUE) {

  ##############################
  # testing parameters
  #DNAstring = BSgenome.Btaurus.UCSC.bosTau9::BSgenome.Btaurus.UCSC.bosTau9$chr1[1:2534110]; gzip = TRUE ; workdir="/media/nmary/DONNEES/bureau_nico/Recherche/R/NoteBooks/"
  #metadataWT = read.table("~/mnt/genome3D/Nicolas/inSilMut/chr1/step1/metadataWT.tsv", header = TRUE, sep = "\t")
  #metadataMT = read.table("~/mnt/genome3D/Nicolas/inSilMut/chr1/step1/metadataMT.tsv", header = TRUE, sep = "\t")
  ###############################

  #create the working directory if it doesn't exist
  if (!dir.exists(workdir)) {
    dir.create(workdir, recursive = TRUE)
  }

  #create the output if it doesn't exist
  outdir = paste0(workdir, "/fasta/")
  if (!dir.exists(outdir)) {
    dir.create(outdir, recursive = TRUE)
  }

  #split metadataMT by start.window
  metadataMT.lst <- split(metadataMT, metadataMT$start.window)
  message("Writing ", nrow(metadataMT), " MT fasta files...")

  #initiate progress bar
  pb <- progress::progress_bar$new(total = length(metadataMT.lst),
                                   format = "[:bar] :percent in :elapsed",
                                   clear = FALSE, width= 60)
  pb$tick(0)

  for (i in 1:length(metadataMT.lst)) {

    subset_metadataMT = metadataMT.lst[[i]]

    START.MAT = unique(subset_metadataMT$start.mat)
    STOP.MAT = unique(subset_metadataMT$stop.mat)
    START.WINDOW = unique(subset_metadataMT$start.window)

    #get 1Mb DNA sequences from metadataWT
    FASTA = DNAstring[START.MAT:STOP.MAT]

    #get DNA frequencies
    freq = c(metadataWT$A[metadataWT$start.window == START.WINDOW],
             metadataWT$T[metadataWT$start.window == START.WINDOW],
             metadataWT$G[metadataWT$start.window == START.WINDOW],
             metadataWT$C[metadataWT$start.window == START.WINDOW])

    MT.fa.lst = lapply(1:nrow(subset_metadataMT), function(i2) {

      shufDNA(dna.string = FASTA,
                       start = subset_metadataMT$start.mut[i2] - subset_metadataMT$start.mat[i2] + 1, #relative position according to FASTA sequence
                       stop = subset_metadataMT$stop.mut[i2] - subset_metadataMT$start.mat[i2] + 1, #relative position according to FASTA sequence
                       probability = freq)
    }) %>% Biostrings::DNAStringSet()

    names(MT.fa.lst) <- subset_metadataMT$ID

    #save the fasta file
    for (i2 in 1:length(MT.fa.lst)) {

      if (isFALSE(gzip)) {
        #write the fasta file
        Biostrings::writeXStringSet(MT.fa.lst[i2], filepath = paste0(outdir, names(MT.fa.lst[i2]), ".fa"))
      }

      if (isTRUE(gzip)) {
        #write the fasta file
        Biostrings::writeXStringSet(MT.fa.lst[i2], filepath = paste0(outdir, names(MT.fa.lst[i2]), ".fa.gzip"), compress = TRUE)
      }
    }
    pb$tick()
  }

  return(message(nrow(metadataMT), " fasta files saved in ", outdir))
}
